libvast/vast/system/query_processor.hpp

```diff
@@ -0,0 +1,162 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <cstdint>
+
+#include <caf/behavior.hpp>
+#include <caf/fwd.hpp>
+
+#include "vast/fwd.hpp"
+#include "vast/uuid.hpp"
+
+namespace vast::system {
+
+/// A query processor takes (1) a query, (2) a processing step function,
+/// and (3) a predicate for implementing the following state machine.
+///
+/// ```
+///                    +----------------+
+///                    |                |
+///               +--->+      idle      |
+///               |    |                |
+///               |    +-------+--------+
+///               |            |
+///               |            | (run)
+///               |            v
+///               |    +-------+--------+
+///               |    |                |
+///               |    | await query id |
+///               |    |                |
+///               |    +-------+--------+
+///               |            |
+///               |            | (query_id, scheduled, total)
+///               |            |
+///               |            |      +------+
+///               |            |      |      |
+///               |            v      v      | (ids)
+///               |    +-------+------+-+    |
+///               |    |                +----+
+///               |    |  collect hits  |
+///               |    |                +<---+
+///               |    +-------+--------+    |
+///               |            |             |
+///               |            | (done)      |
+///               |            v             |
+///               |       XXXXXXXXXXXX       |
+///               |      XX request  XX      |
+///               +----+XX    more    XX+----+
+///                no    XX   hits?  XX   yes
+///                       XXXXXXXXXXXX
+/// ```
```

Comment:
I prefer the back ticks here to avoid Doxygen swallowing the first spaces. This way, the ascii-Art aligns nicely in code as well as in generated output.